runtime/src/main/java/com/flipkart/flux/dao/StatesDAOImpl.java

```diff
@@ -16,99 +16,111 @@
         query.executeUpdate();
     }
 
     @Override
     @Transactional
-    public void incrementRetryCount(Long stateId, Long stateMachineId) {
+    @SelectDataSource(type = DataSourceType.READ_WRITE, storage = storage.SHARDED)
+    public void incrementRetryCount(String stateMachineId, Long stateId) {
         Query query = currentSession().createQuery("update State set attemptedNoOfRetries = attemptedNoOfRetries + 1 where id = :stateId and stateMachineId = :stateMachineId");
         query.setLong("stateId", stateId);
-        query.setLong("stateMachineId", stateMachineId);
+        query.setString("stateMachineId", stateMachineId);
         query.executeUpdate();
     }
 
+    /**
+     * Query should go to Default Shard , As it is a redriver Task
+     *
+     * @param id
+     * @return
+     */
     @Override
     @Transactional
-    public State findById(Long id) {
+    @SelectDataSource(type = DataSourceType.READ_WRITE, storage = storage.SHARDED)
+    public State findById(String stateMachineId, Long id) {
         return super.findById(State.class, id);
     }
 
-    @Override
+
     @Transactional
-    @SelectDataSource(DataSourceType.READ_ONLY)
-    public List findErroredStates(String stateMachineName, Long fromStateMachineId, Long toStateMachineId) {
+    @SelectDataSource(type = DataSourceType.READ_ONLY, storage = storage.SHARDED)
+    public List findErroredStates(ShardId shardId, String stateMachineName, String fromStateMachineId, String toStateMachineId) {
         Query query = currentSession().createQuery("select state.stateMachineId, state.id, state.status from StateMachine sm join sm.states state " +
-                "where sm.id between :fromStateMachineId and :toStateMachineId and sm.name = :stateMachineName and state.status in ('errored', 'sidelined', 'cancelled')");
+                "where sm.id >= :fromStateMachineId and sm.id  <= :toStateMachineId and sm.name = :stateMachineName and state.status in ('errored', 'sidelined', 'cancelled')");
 
-        query.setLong("fromStateMachineId", fromStateMachineId);
-        query.setLong("toStateMachineId", toStateMachineId);
+        query.setString("fromStateMachineId", fromStateMachineId);
+        query.setString("toStateMachineId", toStateMachineId);
         query.setString("stateMachineName", stateMachineName);
-
         return query.list();
     }
 
     @Override
     @Transactional
-    @SelectDataSource(DataSourceType.READ_ONLY)
-    public List findStatesByStatus(String stateMachineName, Timestamp fromTime, Timestamp toTime, String stateName, List<Status> statuses) {
+    @SelectDataSource(type = DataSourceType.READ_ONLY, storage = storage.SHARDED)
+    public List findStatesByStatus(ShardId shardId, String stateMachineName, Timestamp fromTime, Timestamp toTime, String stateName, List<Status> statuses) {
         Query query;
         String queryString = "select state.stateMachineId, state.id, state.status from StateMachine sm join sm.states state " +
```

Comment:
this query returns wrong results, because the insertion order and timestamps are not related in case of string indexes.

Earlier the ids and timestamps are like
100 - t1
101 - t1
102 - t2
103 - t3

and now
abc - t1
xyz - t2 
def - t3

as random strings are inserted in to DB

will discus f2f and solve this